gce/gce.go

```diff
@@ -793,6 +823,58 @@ func (s *gceOps) waitForAttach(
 	return devicePath.(string), nil
 }
 
+// waitForOpCompletion is a blocking function that can be used to check the status
+// of any gce service call. It checks the status of the operation object passed
+// in as argument. It will keep on retrying until
+// 1. gce service returns that the operation has been completed
+// 2. the retry timeout is hit
+// this code has been inspired from kubernetes cloudprovider for gce
+// k8s.io/kubernetes/pkg/cloudprovider/providers/gce/cloud
+func (s *gceOps) waitForOpCompletion(
+	cloudopsOperationName string,
+	opZone string,
+	operation *compute.Operation,
+) error {
+	_, gceOpErr := task.DoRetryWithTimeout(
+		func() (interface{}, bool, error) {
+			// get the status of the operation
+			op, err := s.computeService.ZoneOperations.Get(s.inst.project, opZone, operation.Name).Do()
+			if err != nil {
+				// failed to get operation status
+				// check again later
+				if gErr, ok := err.(*googleapi.Error); ok {
+					if gErr.Code == int(404) {
+						// operation does not exist
+						return nil, false, nil
+					}
+				}
+				return nil, true, fmt.Errorf("failed to query gce operation %v for %v: %v", operation.Name, cloudopsOperationName, err)
+			}
+
+			if op == nil || op.Status != doneStatus {
```

Comment:
Even for failed operations, the status is DONE?

Reply:
Yes that operation status is set to DONE and the error is updated if any.